src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/ExpressionHelper.cs

```diff
@@ -37,86 +38,158 @@ public static string GetExpressionText(LambdaExpression expression, ExpressionTe
                 return expressionText;
             }
 
-            var containsIndexers = false;
+            // Determine size of string needed and number of segments.
+            var lastIsModel = false;
+            var length = 0;
             var part = expression.Body;
-
-            // Builder to concatenate the names for property/field accessors within an expression to create a string.
-            var builder = new StringBuilder();
-
+            var segmentCount = 0;
+            var trailingMemberExpressions = 0;
             while (part != null)
             {
-                if (part.NodeType == ExpressionType.Call)
+                switch (part.NodeType)
                 {
-                    containsIndexers = true;
-                    var methodExpression = (MethodCallExpression)part;
-                    if (!IsSingleArgumentIndexer(methodExpression))
-                    {
-                        // Unsupported.
+                    case ExpressionType.Call:
+                        var methodExpression = (MethodCallExpression)part;
+                        if (IsSingleArgumentIndexer(methodExpression))
+                        {
+                            lastIsModel = false;
+                            length += 4;    // allow room for [99]
```

Comment:
Don't want to determine the actual indexer value up front because that would need to be stored somewhere and it's certainly possible for the expression to be `m => m.Collection[a][b][c]`. So I'm allowing room for an integer indexer that's `99` or less. If it turns out the value is `100` or it's a dictionary w/ long `string` keys, the `StringBuilder` size will double but likely just once (rather than starting at `16` bytes and being resized multiple times). Either way, it's an unusual case where resizing occurs.